compiler/rustc_hir/src/hir.rs

```diff
@@ -647,6 +662,35 @@ impl<'hir> Generics<'hir> {
         )
     }
 
+    /// Returns bounds span for suggestions.
+    /// If the span including lifetime bound needs parentheses, it returns a tuple of a span to be surrounded by parentheses and true.
+    /// e.g. `dyn Future<Output = ()> + 'static` needs parentheses `(dyn Future<Output = ()>) + 'static`
+    pub fn bounds_span_for_suggestions_with_parentheses(
+        &self,
+        param_def_id: LocalDefId,
+    ) -> Option<(Span, bool)> {
+        self.bounds_for_param(param_def_id).flat_map(|bp| bp.bounds.iter().rev()).find_map(
+            |bound| {
+                let span_for_parentheses = bound.get_inner_ty().and_then(|ty| {
```

Comment:
I tried to call `tcx.is_fn_trait` but `Generics` doesn't have `tcx`.
I also tried to pass `tcx` via an argument but `rustc_hir` doesn't have reference to `rustc_middle` which has `TyCtxt`.
Can I add a reference to `rustc_middle` or is there more good solution?

Reply:
Oops, I thought we were in `rustc_middle` ^^' In this case we should just check if the args are parenthesized via `segment.args().parenthesized == GenericArgsParentheses::ParenSugar` which is more general (e.g., it also accounts for `async Fn()` etc.).

Reply:
> In this case we should just check if the args are parenthesized via `segment.args().parenthesized == GenericArgsParentheses::ParenSugar` which is more general (e.g., it also accounts for async Fn() etc.).

Sorry, I couldn't understand why is this condition needed.
My understanding is that it is needed to avoid `async Fn() + 'static`, is it right?
I would appreciate if you could give me an example to use this.

Reply:
The check `segment.args().parenthesized == ParenSugar` should prevent us from adding unnecessary (or even incorrect? dunno) parentheses in cases like `impl SomeTrait<Output = dyn Future<Output = ()>>`.

Reply:
Thank you! I understood and fixed.